src/util/simplify_expr_pointer.cpp

```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
   return true;
 }
 
+static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
+{
+  exprt result = op;
+  while(result.id() == ID_typecast)
+  {
+    result = result.op0();
+  }
+  POSTCONDITION(result.id() == ID_address_of);
+  return result;
+}
+
+// returns true if op is a symbol or an expression like x.y[constant].z
+static bool simplify_inequality_is_simple_structure_expression(const exprt &op)
+{
+  exprt expr = op;
+  while(expr.id() == ID_member || expr.id() == ID_index)
+  {
+    if(expr.id() == ID_member)
+    {
+      expr = to_member_expr(expr).compound();
+    }
+    if(expr.id() == ID_index)
+    {
+      const index_exprt index = to_index_expr(expr);
+      if(!index.index().is_constant())
+      {
+        return false;
+      }
+      expr = index.array();
+    }
+  }
+  return expr.id() == ID_symbol;
+}
+
 bool simplify_exprt::simplify_inequality_address_of(exprt &expr)
```

Comment:
Clearly this was misnamed to begin with, but suggest changing to `simplify_equality_or_inequality...` since the last line makes it clear we're handling both `ID_equal` and `ID_notequal` here.